### PR TITLE
fix: prevent ledger was deactivated on balanced when user login the same wallet address on another platform

### PR DESCRIPTION
**Pull request by shallcdn**
### Step to reproduce:

1.    Login Balanced by Ledger wallet with address hx….01

2.    Login OMM [App | Omm](https://app.omm.finance/#/home)  with the same wallet address as above hx….01

3. Try to perform a transaction on Balanced,

**Actual**: Nothing happened
**Expectation**: Work with any transaction

### Solution:

Renew transport with any transaction need to confirm on Ledger

#932

### CODE REVIEW

**Review by lijac — COMMENTED** [review]
Thank you!
I couldn't test with a ledger, does this change have a direct impact on the issue?
`changeShouldLedgerSign(true);` => `await changeShouldLedgerSign(true);`
I'm asking since I haven't seen using `await` when calling a hook before.

**Comment by shallcdn** [comment]
> Thank you! I couldn't test with a ledger, does this change have a direct impact on the issue? `changeShouldLedgerSign(true);` => `await changeShouldLedgerSign(true);` I'm asking since I haven't seen using `await` when calling a hook before.

Hi @lijac, I think it doesn't impact the issue directly. However, when I found the place to add this block of code [ (initialize a new transport for ledger wallet) ](https://github.com/balancednetwork/balanced-network-interface/pull/943/files#diff-3abfa5f69436e97db717daa3bbb76e71e0308c030703d4c4604dd9e85f633d10L49-R54) to fix this issue, I see this place, I can add the codes as above, and don't need to modify a lot.

Do you have any ideas? Please advise me.

File: src/store/application/hooks.ts [hunk 1]
```diff
@@ -46,7 +48,10 @@ export function useCurrentLedgerAddressPage(): AppState['application']['currentL
 export function useChangeShouldLedgerSign(): (shouldLedgerSign: boolean) => void {
```
Comment (lijac) [inline src/store/application/hooks.ts:48]: The return value will be `Promise<void>`

**Review by lijac — COMMENTED** [review]
We can just call `changeShouldLedgerSign(true);` without `await`.

File: src/packages/BalancedJs/index.ts [hunk 2]
```diff
@@ -148,4 +149,22 @@ export class BalancedJs {
   getContract(address: string): IRC2 {
     return new IRC2(this.contractSettings, address);
   }
+
+  async initialiseTransport(): Promise<void> {
+    debugger;
```
[the remark this answers is not in the document]
Comment (lijac) [reply src/packages/BalancedJs/index.ts:154]: Let's use `initializeTransport` following the US style.

**Comment by shallcdn** [comment]
> We can just call `changeShouldLedgerSign(true);` without `await`.

Because function `initializeTransport` contain this block code
```
const transport = await TransportWebHID.create();
```
 So that if we don't use `await` for `changeShouldLedgerSign(true);` the transport of Ledger cannot renew before calling to create a new transaction

**Comment by lijac** [comment]
> So that if we don't use `await` for `changeShouldLedgerSign(true);` the transport of Ledger cannot renew before calling to create a new transaction

Well, do we have to ensure ledger transport renewal is complete before moving to the next operation? Can you check that again with the ledger?

File: src/packages/BalancedJs/index.ts [hunk 3]
```diff
@@ -148,4 +149,21 @@ export class BalancedJs {
...
+      this.contractSettings.ledgerSettings.transport.close();
+      this.contractSettings.ledgerSettings = {} as LedgerSettings;
+      const transport = await TransportWebHID.create();
+      if (transport.setDebugMode) {
+        transport.setDebugMode(false);
+      }
+      this.inject({
+        legerSettings: {
+          transport,
+          path: getLedgerAddressPath(-1),
```
Comment (lijac) [inline src/packages/BalancedJs/index.ts:164]: Can you leave a comment on what `-1` means?

**Comment by shallcdn** [comment]
> > So that if we don't use `await` for `changeShouldLedgerSign(true);` the transport of Ledger cannot renew before calling to create a new transaction
> 
> Well, do we have to ensure ledger transport renewal is complete before moving to the next operation? Can you check that again with the ledger?

Yes, I have checked it. If I remove `await`, the Iconex/Hanna wallet will pop up instead ledger

**Comment by BennyOptions** [comment]
I tested with ledger and it seems to work @shallcdn

**Comment by lijac** [comment]
Thank you! Let me think about possible improvements.

File: src/packages/BalancedJs/index.ts (hunk 3 above)
Comment (lijac) [reply src/packages/BalancedJs/index.ts]: Updated

**Comment by shallcdn** [comment]
> I tested with ledger and it seems to work @shallcdn

Thank you!

**Comment by shallcdn** [comment]
> Thank you! Let me think about possible improvements.

Thank you

**Comment by octocat512** [comment]
@shallcdn please resolve the conflicts.

**Review by lijac — APPROVED** [review]

**Pull request merged**